File: apps/backend-gateway/src/application/schemas/gate_schemas.py
"""
StadiumOS AI — Gate API Pydantic Schemas.

Validation and serialization DTO schemas for Gate API endpoints.
"""

from uuid import UUID
from pydantic import BaseModel, Field

from src.domain.enums import GateStatus


class GateCreate(BaseModel):
    """Payload to create a new gate."""

    gate_code: str = Field(..., min_length=2, max_length=50, description="Gate identification code (e.g. 'GATE_1A').")
    latitude: float = Field(..., ge=-90.0, le=90.0)
    longitude: float = Field(..., ge=-180.0, le=180.0)
    is_bidirectional: bool = Field(default=True)


class GateStatusUpdate(BaseModel):
    """Payload to update a gate's operational status."""

    status: GateStatus = Field(..., description="Target status (OPEN, CLOSED, RESTRICTED, CONGESTED, MAINTENANCE).")


class GateResponse(BaseModel):
    """Serialized gate data."""

    id: UUID
    sector_id: UUID
    gate_code: str
    status: GateStatus
    latitude: float
    longitude: float
    is_bidirectional: bool

    class Config:
        from_attributes = True

File: apps/backend-gateway/src/application/schemas/gateSchemas.ts
/**
 * StadiumOS AI — Gate Validation Schemas.
 * 
 * Defines Zod schemas for gate creation and state modifications.
 */

import { z } from "zod";
import { GateStatus } from "../../domain/enums";

export const GateCreateSchema = z.object({
  gate_code: z.string().min(2).max(50),
  latitude: z.number().min(-90.0).max(90.0),
  longitude: z.number().min(-180.0).max(180.0),
  status: z.nativeEnum(GateStatus).default(GateStatus.CLOSED),
  is_bidirectional: z.boolean().default(true)
});

export type GateCreate = z.infer<typeof GateCreateSchema>;

export const GateStatusUpdateSchema = z.object({
  status: z.nativeEnum(GateStatus),
  reason: z.string().min(10, "A detailed operational reason (minimum 10 characters) is required.")
});

export type GateStatusUpdate = z.infer<typeof GateStatusUpdateSchema>;
